core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala

```diff
@@ -92,8 +95,15 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
   /** The checksum for all the blocks. */
   private var checksums: Array[Int] = _
 
-  override protected def getValue() = {
-    _value
+  override protected def getValue() = synchronized {
```

Comment:
I noticed that below, the readBroadcastBlock() method is synchronized on the companion object, which makes me nervous. However a lazy val is also implemented with `this.synchronized`, so I suspect this is fine. We pay the cost of synchronization on every call now, but I think that is OK here, as a simple flag won't help us figure out whether the SoftReference has been cleared.

Reply:
I think we can remove the `TorrentBroadcast.synchronized` in `readBroadcastBlock`, since we're already synchronizing in its only caller? Though I'm not sure why it was necessary in the first place, as `readBroadcastBlock` should only have been called once before this PR.

Regardless, I agree that the perf hit should be ok. Let me know if you want any of this changed.

Reply:
Hm, that `TorrentBroadcast.synchronized` is from a very old version of the code in 2013, when more things used that lock. I'm pretty certain it's obsolete. However this code accesses broadcastCache and that needs synchronization. (It's kind of unfortunate where this object is). I think we could actually improve this by locking on broadcastCache for basically the whole block. I think that's a safe improvement as nothing else uses broadcastCache, nothing else is synchronized here, and should behave just the same.

Reply:
Do you mean switching `TorrentBroadcast.synchronized` to `broadcastCache.synchronized` inside `readBroadcastBlock`, or changing `this.synchronized` to `broadcastCache.synchronized` inside `getValue()` (and getting rid of the lock in `readBroadcastBlock`?

Reply:
We need two locks, one to protect the local reference to the block, and one to protect the shared cache object. I was thinking of the former.

Reply:
done